File: rust/src/day06.rs
use std::collections;

fn startpacket<const WINDOW_SIZE: usize>(datastream: &str) -> usize {
    let start_index = datastream
        .as_bytes()
        .array_windows::<WINDOW_SIZE>()
        .enumerate()
        .find_map(|(index, chunk)| {
            let mut set = collections::HashSet::new();
            let all_distinct = chunk.iter().all(|&char| set.insert(char));
            if all_distinct {
                Some(index)
            } else {
                None
            }
        })
        .unwrap();

    start_index + WINDOW_SIZE
}

pub fn part1() -> usize {
    const WINDOW_SIZE: usize = 4;
    let input = include_str!("data/day06.txt");
    startpacket::<WINDOW_SIZE>(input)
}

pub fn part2() -> usize {
    const WINDOW_SIZE: usize = 14;
    let input = include_str!("data/day06.txt");
    startpacket::<WINDOW_SIZE>(input)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_startpacket() {
        const WINDOW_SIZE: usize = 4;
        let data = [
            "mjqjpqmgbljsphdztnvjfqwrcgsmlb",
            "bvwbjplbgvbhsrlpgdmjqwftvncz",
            "nppdvjthqldpwncqszvftbrmjlhg",
            "nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg",
            "zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw",
        ];
        for (index, &packet) in data.iter().enumerate() {
            match index {
                0 => assert_eq!(startpacket::<WINDOW_SIZE>(packet), 7),
                1 => assert_eq!(startpacket::<WINDOW_SIZE>(packet), 5),
                2 => assert_eq!(startpacket::<WINDOW_SIZE>(packet), 6),
                3 => assert_eq!(startpacket::<WINDOW_SIZE>(packet), 10),
                4 => assert_eq!(startpacket::<WINDOW_SIZE>(packet), 11),
                _ => unreachable!(),
            }
        }
    }

    #[test]
    fn test_part1() {
        assert_eq!(part1(), 1987);
    }

    #[test]
    fn test_part2() {
        assert_eq!(part2(), 3059);
    }
}

File: zig/src/day06.zig
const std = @import("std");
const fba = @import("alloc.zig");

const testing = std.testing;
const mem = std.mem;

const Set = std.AutoArrayHashMap(u8, void);

fn startpacket(datastream: []const u8, chunk_size: usize) usize {
    var win_iter = mem.window(u8, datastream, chunk_size, 1);
    var chunk_count: usize = 0;
    while (win_iter.next()) |chunk| : (chunk_count += 1) {
        var set = Set.init(fba.arena);
        set.ensureTotalCapacity(chunk_size) catch unreachable;
        defer set.deinit();

        for (chunk, 1..) |value, index| {
            const entry = set.getOrPutAssumeCapacity(value);
            if (entry.found_existing) {
                break;
            } else {
                if (index == chunk_size) {
                    return chunk_count + chunk_size;
                }
                entry.key_ptr.* = value;
            }
        }
    }
    unreachable;
}

test startpacket {
    {
        const marker_size = 4;
        const data = [_][]const u8{
            "mjqjpqmgbljsphdztnvjfqwrcgsmlb",
            "bvwbjplbgvbhsrlpgdmjqwftvncz",
            "nppdvjthqldpwncqszvftbrmjlhg",
            "nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg",
            "zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw",
        };
        for (data, 0..) |packet, index| {
            switch (index) {
                0 => try testing.expectEqual(startpacket(packet, marker_size), 7),
                1 => try testing.expectEqual(startpacket(packet, marker_size), 5),
                2 => try testing.expectEqual(startpacket(packet, marker_size), 6),
                3 => try testing.expectEqual(startpacket(packet, marker_size), 10),
                4 => try testing.expectEqual(startpacket(packet, marker_size), 11),
                else => unreachable,
            }
        }
    }
    {
        const marker_size = 14;
        const data = [_][]const u8{
            "mjqjpqmgbljsphdztnvjfqwrcgsmlb",
            "bvwbjplbgvbhsrlpgdmjqwftvncz",
            "nppdvjthqldpwncqszvftbrmjlhg",
            "nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg",
            "zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw",
        };
        for (data, 0..) |packet, index| {
            switch (index) {
                0 => try testing.expectEqual(startpacket(packet, marker_size), 19),
                1 => try testing.expectEqual(startpacket(packet, marker_size), 23),
                2 => try testing.expectEqual(startpacket(packet, marker_size), 23),
                3 => try testing.expectEqual(startpacket(packet, marker_size), 29),
                4 => try testing.expectEqual(startpacket(packet, marker_size), 26),
                else => unreachable,
            }
        }
    }
}

pub fn part1() usize {
    const input = @embedFile("data/day06.txt");
    const marker_size = 4;
    return startpacket(input, marker_size);
}

test part1 {
    try testing.expectEqual(part1(), 1987);
}

pub fn part2() usize {
    const input = @embedFile("data/day06.txt");
    const marker_size = 14;
    return startpacket(input, marker_size);
}

test part2 {
    try testing.expectEqual(part2(), 3059);
}
